src/core/extractor.py

```diff
@@ -121,9 +183,21 @@ def decorator(cls: Type["MetadataBase"]) -> Type["MetadataBase"]:
 
     def __init__(self, logger: Optional[Logger] = None):
         self._logger = logger or logging.getLogger(__name__)
-        # todo: change class setupt to factory pattern to avoid having to manually calling setup after construction
-        #       see https://stackoverflow.com/a/33134213/2160256
-        # self.setup()
+
+    async def setup(self) -> None:
+        """Fetch tag lists from configured urls."""
+
+        if len(self.urls) > 0:
+            assert self.tag_list == [], "Cannot specify both urls and tag list"
```

Comment:
What is the use of an assert in production code?

Reply:
Even in production asserts are enabled (unless explicitly disabled via `-NDEBUG` or something similar). 
If there is an empty list, we want the app to crash, because its a bug :)

Reply:
So basically, this is a if(...), raise Exception? Would raising an exception not be ... cleaner/more explicit?

Reply:
yep, changed to exception

Reply:
Thank you!